### DISCUSSION TITLE
Loom Based Executor question

### DISCUSSION THREAD

**Discussion by IgorDorokhov** [zio/zio#9375] [Q&A] [answered]
Hi everyone,
I have a question regarding Runtime.enableLoomBasedExecutor.
I think loom based executor can be a good option in a scenario where internal dependencies of a ZIO project are written in java (21 or up) and use virtual threads themselves.

After more reading I found following release notes https://github.com/zio/zio/releases/tag/v2.1.0-RC2 where following quote is made: "_We also added an optional executor based on Loom that you can use if your Java version is 21 or higher. It looks like our default ZScheduler is faster when forking a lot of fibers, but depending on your use case, you might want to try this new executor using the Runtime.enableLoomBasedExecutor aspect._"

My question is: why loom based executor is slower than ZScheduler when forking a lot of fibers? What exactly "forking a lot of fibers" mean? Does it mean actual usage of ".fork" or it means in general spawning a lot of fibers? But ZIO is based on fibers and it does spawn a lot of fibers by itself.
And another question is when would you recommend to use loom based executor instead of ZScheduler?

As per my understanding benefits of loom based executor are following:
- ZIO application will utilize virtual threads provided by Project Loom instead of traditional thread pools this allows for handling a large number of concurrent tasks with better resource utilization
- Virtual threads are particularly beneficial for applications that spend a lot of time waiting on I/O operations
- They reduce the overhead associated with traditional threads, such as context switching and memory consumption
- More efficient handling of a large number of concurrent tasks

Does loom based executor can handle more load (throughput) than Zscheduler but in expense of increased latency?

Can it be slower due to higher load on platform threads? Meaning that a large number of fibers can be created per virtual thread, and large amount of virtual threads can be created per platform threads - and that puts a lot of load on platform threads? Is that the case?

Thank you!

**Comment by IgorDorokhov** [comment]
@ghostdogpr Hi, could you please answer the question above?

**Comment by ghostdogpr** [comment, marked as answer]
The benchmark where it was slower is this one: https://github.com/zio/zio/blob/series/2.x/benchmarks/src/main/scala/zio/ForkJoinBenchmark.scala

You can see more details in the PR description here: https://github.com/zio/zio/pull/8745

But that benchmark is not really realistic of actual production usage, so we encourage you to test it yourself. I've actually heard from a couple people that it was performing well. Note that Loom is not perfect: synchronized blocks will cause pinning of the carrier thread (blocking the main thread, not the virtual thread). I think Java 24 might include https://openjdk.org/jeps/491 and solve it.
